framework/src/mesh/MooseMesh.C

```diff
@@ -865,6 +866,11 @@ MooseMesh::cacheInfo()
       std::set<BoundaryID> & subdomain_set = _subdomain_boundary_ids[subdomain_id];
 
       subdomain_set.insert(boundaryids.begin(), boundaryids.end());
+
```

Comment:
I am a little puzzled how this function works. `getMesh().element_ptr_range()` looks to me all the elements on local processors. With distributed mesh, the range does not include elements on other processors except the ghost elements. Thus supposedly, we should have a communication at the end of this function to set the union of `_subdomain_boundary_ids` and `_neighbor_subdomain_boundary_ids`. Did I miss something?

Reply:
Oh, we did see distributed mesh failures.

Reply:
I think the proper way is to use `_mesh_subdomains` to create all entries. And after the element loop, we call `_communicator.set_union()` on all entries.

`_block_node_list` is a little problematic unless the relation manager put all elements surrounding the nodes as ghost elements when the node is on processor interfaces. It is irrelevant to this PR though.

Reply:
So, I put a `_communicator.set_union()` at the end of this function, and now there is a csvdiff. After some debugging, I realized there must be something wrong with the `getBoundaryIDs(elem, side)` function. To demonstrate, I put this at the end of `cacheInfo()`:

```language=cpp
  std::this_thread::sleep_for(std::chrono::milliseconds(processor_id()));
  std::cout << "processor : " << processor_id() << std::endl;
  for (const auto & blk_id : _mesh_subdomains)
  {
    std::cout << "    subdomain " << blk_id << ":";
    for (const auto & itt : _subdomain_boundary_ids[blk_id])
      std::cout << " " << itt;
    std::cout << std::endl;
  }
  std::this_thread::sleep_for(std::chrono::milliseconds(10));
```
which outputs what each processor has for `_subdomain_boundary_ids`. 

This is the result in serial:
```
processor : 0
    subdomain 0: 4
    subdomain 1: 2 3
    subdomain 2: 2
    subdomain 3: 1 2
    subdomain 4: 1
    subdomain 5: 0 1
    subdomain 6: 0
    subdomain 7: 0 3
    subdomain 8: 3
```
which is correct. Now the result distributed on 2 processors:
```
processor : 0
    subdomain 0: 2 3 4
    subdomain 1:
    subdomain 2: 2
    subdomain 3:
    subdomain 4:
    subdomain 5: 3
    subdomain 6: 0
    subdomain 7: 0 3
    subdomain 8:
processor : 1
    subdomain 0:
    subdomain 1: 1
    subdomain 2: 2
    subdomain 3: 1 2
    subdomain 4: 0 1
    subdomain 5:
    subdomain 6: 0
    subdomain 7:
    subdomain 8:
```
which is completely wrong.

Am I missing something? I am only testing what was already in MooseMesh, so I'm not considering neighbor elements at all.

Reply:
Can you push your changes here? `getBoundaryIDs(elem, side)` is directly calling libMesh API, I double there is mistakes there.

Reply:
FYI - I know it's mostly unrelated, but `SerializerGuard` is awesome for this kind of debugging

```
{
  SerializerGuard sg(_communicator);
  for (const auto & blk_id : _mesh_subdomains)
  {
    std::cout << "    subdomain " << blk_id << ":";
    for (const auto & itt : _subdomain_boundary_ids[blk_id])
      std::cout << " " << itt;
    std::cout << std::endl;
  }
}
```
...without the sleep madness :-)

Reply:
Note: `set_union` needs to be called on each element of the map independently.

Reply:
> Can you push your changes here? `getBoundaryIDs(elem, side)` is directly calling libMesh API, I double there is mistakes there.

You can just copy my debugging code after line 875 in MooseMesh.C.

Reply:
I also pushed the debugging commit.

Reply:
Try it with `allow_renumbering = false` in the `Mesh` block.

Elements may have different IDs with distributed mesh which will change the order in which they're accessed.

Reply:
..err. Your output wasn't using `getBoundaryIDs()` so definitely disregard.

Reply:
Crud. Disregard again. Your issue is probably renumbering. The `ElementSubdomainIDGenerator` assigns the subdomains vs IDs. The IDs are different in distributed/replicated. Therefore - the subdomains are different.

Reply:
oh?! Then sorry Zach, let us use your original input.

Reply:
I am surprised if this is the case, because I remember Roy worked on that mesh generator in the past to make sure it works fine for distributed mesh. I might be wrong.

Reply:
I see bunch of tests in MOOSE using this generator and those tests seem working fine with distributed mesh. I am puzzled.

Reply:
My old input using `SubdomainBoundingBoxGenerator` works with the `set_union`. Looks like the issue is with  `ElementSubdomainIDGenerator`.

Reply:
I'm not sure how to fix the issue with `ElementSubdomainIDGenerator`, so I'll just revert to using my old input. Then create an issue once this is merged.

Reply:
If you set `allow_renumbering = false` in the mesh block, your input with `ElementSubdomainIDGenerator` will work

Reply:
There really doesn't need to be an issue with the generator either. This all stems from the fact that we allow things to be renumbered with distributed mesh due to partitioning.